Framework/PythonInterface/plugins/algorithms/ConvertQtoHKLMDHisto.py

```diff
@@ -0,0 +1,168 @@
+
+        self.declareProperty(IntArrayProperty("Bins", [301, 301, 301],
+                                              direction=Direction.Input),
+                             "Number of bins to use for each dimension, Enter it as a"
+                             "comma-separated list of integers.")
+
+        self.declareProperty(IMDHistoWorkspaceProperty("OutputWorkspace", "",
+                                                       optional=PropertyMode.Mandatory,
+                                                       direction=Direction.Output),
+                             doc="Output MDWorkspace in Q-space, name is prefix if multiple input files were provided.")
+
+    def validateInputs(self):
+        issues = dict()
+
+        input_ws = self.getProperty("InputWorkspace").value
+
+        if input_ws.getSpecialCoordinateSystem().name != "QSample":
+            issues["InputWorkspace"] = "Input workspace expected to be in QSample, " \
+                                       "workspace is in '{}'".format(input_ws.getSpecialCoordinateSystem().name)
+
+        ndims = input_ws.getNumDims()
+        if ndims != 3:
+            issues["InputWorkspace"] = "Input workspace needs 3 dimensions, it has {} dimensions.".format(ndims)
+
+        # Check that extents and bins were provided for every dimension
+        extents = self.getProperty("Extents").value
+        bins = self.getProperty("Bins").value
+
+        if len(extents) != ndims * 2:
+            issues["Extents"] = "Expected a min and max value for each " \
+                                "dimension (got {}, expected {}).".format(len(extents), ndims*2)
+
+        if len(bins) != ndims:
+            issues["Bins"] = "Expected a number of bins for each dimension."
+
+        if self.getProperty("FindUBFromPeaks").value:
+            peak_ws = self.getProperty("PeaksWorkspace")
+            if peak_ws.isDefault:
+                issues["PeaksWorkspace"] = "A peaks workspace must be supplied."
+            elif not mtd.doesExist(self.getPropertyValue("PeaksWorkspace")):
+                issues["PeaksWorkspace"] = "Provided peaks workspace does not exist in the ADS."
+
+            tmatrix = self.getProperty("Transformation")
+            if not tmatrix.isDefault and len(tmatrix.value) != 9:
+                issues["Transformation"] = "Invalid transformation matrix, there should be 9 values comma separated"
+        else:
+            # Check that the workspace has a UB matrix
+            if not (input_ws.getNumExperimentInfo() > 0 and input_ws.getExperimentInfo(0).sample().hasOrientedLattice()):
+                issues["InputWorkspace"] = "Could not find a UB matrix in this workspace."
+
+        return issues
+
+    def PyExec(self):
+        input_ws = self.getProperty("InputWorkspace").value
+
+        extents = self.getProperty("Extents").value
+        bins = self.getProperty("Bins").value
+
+        if self.getProperty("FindUBFromPeaks").value:
```

Comment:
You could simplfy this by removing the `FindUBFromPeaks` option and just have it that if a `PeaksWorkspace` is provided use the UB from it.